### ISSUE TITLE
test-compress fails if lz4 tools aren't installed

### ISSUE THREAD

**Issue by jhxie** [systemd/systemd#7441]
### Submission type

 - Bug report

### systemd version the issue has been seen with

build version - master (8a38aac)
packaged distribution version - 234

### Used distribution

Ubuntu 17.10 x86_64

### In case of bug report: Expected behaviour you didn't see

test-path-util passes.

### In case of bug report: Unexpected behaviour you saw

```
The output from the failed tests:

136/236 test-path-util                          FAIL     0.11 s

--- command ---
SYSTEMD_KBD_MODEL_MAP='/home/jiahui/projects/systemd/src/locale/kbd-model-map' SYSTEMD_LANGUAGE_FALLBACK_MAP='/home/jiahui/projects/systemd/src/locale/language-fallback-map' PATH='/home/jiahui/projects/systemd/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games' /home/jiahui/projects/systemd/build/test-path-util
--- stdout ---
/bin/sh
/home/jiahui/projects/systemd/build/test-path-util
/bin/sh
--- stderr ---
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
---/a/b/c/d---
---/a/b/c---
---/a/b---
---/a---
------
---/a/b/c---
---/a/b---
---/a---
------
Assertion 'fsck_exists("minix") == 1' failed at ../src/test/test-path-util.c:254, function test_fsck_exists(). Aborting.
-------
```

### In case of bug report: Steps to reproduce the problem
After following the build instructions listed in the [HACKING](https://github.com/systemd/systemd/blob/master/HACKING) file up to the step:
```bash
ninja -C build test
```
I have double checked whether the `fsck.minix` program is properly installed with the following:
```bash
which fsck.minix
/sbin/fsck.minix
```
Thank you for looking into this issue.

**Comment by evverx** [comment]
On Ubuntu `-Dsplit-usr=true` should be used to make the test work. Could you show the output of `meson configure build/ | grep -i split-usr`?

**Comment by jhxie** [comment]
I removed the `build` directory entirely and build it as suggested with extra flag attached:
```bash
rm -rf build
meson build/ -Dsplit-usr=true
```
The output of the command you have suggested is as follows (extra space removed):
```bash
split-usr true [True, False] assume that /bin, /sbin aren't symlinks into /usr
```

**Comment by jhxie** [comment]
This time the stated test failure is gone; however, a different test failure shows up:
```
The output from the failed tests:

164/240 test-compress                           FAIL     0.37 s

--- command ---
SYSTEMD_KBD_MODEL_MAP='/home/jiahui/projects/systemd/src/locale/kbd-model-map' SYSTEMD_LANGUAGE_FALLBACK_MAP='/home/jiahui/projects/systemd/src/locale/language-fallback-map' PATH='/home/jiahui/projects/systemd/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games' /home/jiahui/projects/systemd/build/test-compress
--- stdout ---
Binary files /home/jiahui/projects/systemd/build/test-compress and - differ
--- stderr ---
/* testing XZ text blob compression/decompression */
/* testing XZ random blob compression/decompression */
compression failed: No buffer space available
/* testing decompress_startswith with XZ on text text */
/* testing decompress_startswith with XZ on random text */
compression failed: No buffer space available
/* testing decompress_startswith with XZ on HUGE=xxxxxxxxxxxxxxx text */
compression failed: No buffer space available
/* testing XZ compression */
/* create source from /home/jiahui/projects/systemd/build/test-compress */
/* test compression */
XZ compression finished (31792 -> 9804 bytes, 30.8%)
/* test decompression */
XZ decompression finished (9804 -> 31792 bytes, 324.3%)
/* test faulty decompression */
Decompression failed: code 7
/* testing LZ4 text blob compression/decompression */
/* testing LZ4 random blob compression/decompression */
compression failed: No buffer space available
/* testing decompress_startswith with LZ4 on text text */
/* testing decompress_startswith with LZ4 on random text */
compression failed: No buffer space available
/* testing decompress_startswith with LZ4 on HUGE=xxxxxxxxxxxxxxx text */
compression failed: No buffer space available
/* testing LZ4 compression */
/* create source from /home/jiahui/projects/systemd/build/test-compress */
/* test compression */
Buffer size is 851984 bytes, header size 7 bytes.
LZ4 compression finished (31792 -> 17129 bytes, 53.9%)
sh: 1: lz4cat: not found
Assertion 'system(cmd) == 0' failed at ../src/journal/test-compress.c:178, function test_compress_stream(). Aborting.
-------

Full log written to /home/jiahui/projects/systemd/build/meson-logs/testlog.txt
FAILED: meson-test
/usr/bin/python3 -u /usr/share/meson/meson test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.
```
By the way, the Ubuntu distribution is installed in a virtual machine (Hyper-V, specifically); but I don't think this has anything to do with this test failure.

**Comment by evverx** [comment]
The output indicates that `lz4cat` has not been installed, so `apt-get install liblz4-tool` should fix the issue.

I think it would be great if `README` mentioned that `l4cat` is needed if `HAVE_LZ4` is set. Could you send a PR?

**Comment by jhxie** [comment]
Sorry I did not spot that warning, silly me.
For the current version [README](https://github.com/systemd/systemd/blob/master/README), it explicitly mentions "liblz4 >= 119 (optional)"; but unfortunately I could not locate any documentation regarding the `HAVE_LZ4` macro and meson `want_lz4` option. I think by default meson would set this option to `false` if `lz4cat` program can not be located (on line 969 of [meson.build](https://github.com/systemd/systemd/blob/master/meson.build) file)?

**Comment by evverx** [comment]
http://mesonbuild.com/Reference-manual.html#dependency is used to check whether a library (in this case, `liblz4`) is installed on your system. The presence of a certain library doesn't mean that something using this library is installed too.

**Comment by jhxie** [comment]
Thanks for the info on meson.
Any suggestions on where should I make the changes you mentioned in the [README](https://github.com/systemd/systemd/blob/master/README) file? Would the `REQUIREMENTS` section be a good starting point? I did not notice anything similar like `HAVE_XZ` defined in the documentation.

**Comment by jhxie** [comment]
I also noticed the continuous integration system (Semaphore-CI) used by the `systemd` project involves building the project itself on a Debian/Ubuntu distribution with the following command specifically:
```bash
~/.local/bin/meson build -Dtests=unsafe -Dsplit-usr=true -Dslow-tests=true
```
As of now only the `-Dsplit-usr=true` option is explicitly mentioned in the [README](https://github.com/systemd/systemd/blob/master/README) file; running command like
```bash
grep --exclude-dir=build/ -nr 'tests=unsafe' .
```
in the root directory of the project does not yield any result unfortunately.
If I were going to send a PR, should I consider add those two extra flags, too?

**Comment by poettering** [comment]
So, the test-path-util thing is a local configuration error, but the fact that the tests fail with lz4cat missing is something we should fix I figure

**Comment by poettering** [comment]
PR #7455 should fix the test-compress issue for you. Could you have a look please?

**Comment by jhxie** [comment]
I removed the `build` directory entirely in addition to the `liblz4-tool` package as mentioned by Evgeny and applied the PR #7455 to the current master branch; after configure and build process this time the test suite passes perfectly on Ubuntu Server 17.10.
Thank you for resolving this issue.

**Comment by poettering** [comment]
It's not resolved yet as #7455 is not merged yet. Note that github automatically closes the right issue when we merge a PR that fixes it. Hence, I'll reopen this for now, github will then take over the remaining work as soon as #7455 is merged.
